File: FortyTwo/Shared/Models/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FortyTwo.Shared.Models
{
    public class Game
    {
        public Game() { }

        public Game(string name)
        {
            Name = name;
            Id = Guid.NewGuid();
        }

        public Guid Id { get; set; } // TODO: in practice, this prolly shouldn't have a public setter
        public string Name { get; set; }
        public string FirstActionBy { get; set; }
        public Bid? Bid { get; set; }
        public string BiddingPlayerId { get; set; }
        public Suit? Trump { get; set; }
        public string CurrentPlayerId { get; set; }

        // TODO: consider changing this to be a dictionary
        public List<Hand> Hands { get; set; } = new List<Hand>(); // this shouldn't be fully exposed
        public Trick CurrentTrick { get; set; } = new Trick();
        public List<Trick> Tricks { get; set; } = new List<Trick>();
        public int? Value
        {
            get 
            {
                if (BiddingPlayerId == null) return null;

                var bid = Hands?.FirstOrDefault(x => x.PlayerId == BiddingPlayerId)?.Bid;
                return bid == null
                    ? (int?)null
                    : (int)bid <= 42 ? 1 : (int)bid / 42; 
            }
        }
        public Teams? WinningTeam {
            get 
            {
                if (BiddingPlayerId == null) return null;

                var biddingTeamId = Hands?.FirstOrDefault(x => x.PlayerId == BiddingPlayerId)?.Team;
                var otherTeamId = Hands?.FirstOrDefault(x => x.Team != biddingTeamId)?.Team;
                var teamPoints = Tricks.GroupBy(t => t.Team).ToDictionary(g => g.Key, g => g.Sum(t => t.Value));

                var adjustedBid = (int)Bid % 42 == 0 ? 42 : (int)Bid;
                return teamPoints.TryGetValue(biddingTeamId, out var biddingTeamPoints) && biddingTeamPoints >= adjustedBid
                    ? biddingTeamId
                    : teamPoints.TryGetValue(otherTeamId, out var otherTeamPoints) && otherTeamPoints > (42 - adjustedBid)
                        ? otherTeamId
                        : null;
            }
        }
    }
}

File: cloudflare/packages/rules/src/game.ts
import { Bid } from './bid';
import { Hand } from './hand';
import { Suit } from './suit';
import { Teams } from './teams';
import { Trick, trickValue } from './trick';

export interface Game {
  id: string;
  name: string;
  firstActionBy: string | null;
  bid: Bid | null;
  biddingPlayerId: string | null;
  trump: Suit | null;
  currentPlayerId: string | null;
  hands: Hand[];
  currentTrick: Trick;
  tricks: Trick[];
}

// Port of C# `Game.Value`.
export function gameValue(g: Game): number | null {
  if (g.biddingPlayerId == null) return null;

  const bid = g.hands.find((x) => x.playerId === g.biddingPlayerId)?.bid;
  return bid == null ? null : bid <= 42 ? 1 : Math.floor(bid / 42);
}

// Port of C# `Game.WinningTeam`. Kept structurally faithful to the original getter
// (same four-way ternary shape) rather than flattened, so it stays auditable line-by-line
// against FortyTwo/Shared/Models/Game.cs.
export function gameWinningTeam(g: Game): Teams | null {
  if (g.biddingPlayerId == null) return null;

  const biddingTeamId = g.hands.find((x) => x.playerId === g.biddingPlayerId)?.team ?? null;
  const otherTeamId = g.hands.find((x) => x.team !== biddingTeamId)?.team ?? null;

  // teamPoints only has an entry for a team once they've won at least one trick
  // (mirrors C#'s `Tricks.GroupBy(t => t.Team).ToDictionary(...)`: a team absent from
  // Tricks is simply absent as a key, not present with value 0).
  const teamPoints = new Map<Teams, number>();
  for (const t of g.tricks) {
    if (t.team == null) continue;
    teamPoints.set(t.team, (teamPoints.get(t.team) ?? 0) + trickValue(t));
  }

  const adjustedBid = (g.bid as number) % 42 === 0 ? 42 : (g.bid as number);

  return g.trump === Suit.Low
    ? biddingTeamId != null && teamPoints.has(biddingTeamId)
      ? otherTeamId
      : g.tricks.length === 7
        ? biddingTeamId
        : null
    : biddingTeamId != null && teamPoints.has(biddingTeamId) && (teamPoints.get(biddingTeamId) as number) >= adjustedBid
      ? biddingTeamId
      : otherTeamId != null && teamPoints.has(otherTeamId) && (teamPoints.get(otherTeamId) as number) > 42 - adjustedBid
        ? otherTeamId
        : null;
}
